lcls_tools/common/image/fit.py

```diff
@@ -0,0 +1,164 @@
+                if bounding_box_penalty > 0 and self.apply_bounding_box_constraint:
+                    fit_results["centroid"] = np.array((np.Nan, np.Nan))
+                    fit_results["rms_sizes"] = np.array((np.Nan, np.Nan))
+
+                    warnings.warn(
+                        f"Beam bounding box is outside the image, penalty value: "
+                        f"{bounding_box_penalty}"
+                    )
+
+            else:
+                fit_results["centroid"] = np.array((np.Nan, np.Nan))
+                fit_results["rms_sizes"] = np.array((np.Nan, np.Nan))
+                fit_results["bb_penalty"] = np.Nan
+
+                warnings.warn(
+                    "Image fits returned Nan values"
+                )
+
+        return fit_results
+
+    def _calculate_bounding_box_penalty(self, image, centroid, sizes):
+        """
+        Calculate bounding box penalty which determines if the beam bounding box is
+        inside or outside the image. Value is < 0 if inside and > 0 if outside /
+        partically outside.
+
+        Note: bounding box penalty function should be continuous to be an effective
+        constraint. See https://www.mdpi.com/2410-390X/7/3/29 for details.
+        """
+        n_stds = self.bounding_box_half_width
+
+        bounding_box_corner_pts = np.array(
+            (
+                centroid - n_stds * sizes,
+                centroid + n_stds * sizes,
+                centroid - n_stds * sizes * np.array((-1, 1)),
+                centroid + n_stds * sizes * np.array((-1, 1)),
+            )
+        )
+
+        if self.image_processor.roi is not None:
+            roi_radius = self.image_processor.roi.radius
+        else:
+            roi_radius = np.min(np.array(image.shape) / 1.5)
+            # TODO: maybe need to change this ^^^
+
+        # This is a check whether or not the beamspot is within
+        # the bounding box.
+        temp = bounding_box_corner_pts - np.array((roi_radius, roi_radius))
+        distances = np.linalg.norm(temp, axis=1)
+        bounding_box_penalty = np.max(distances) - roi_radius
+
+        return bounding_box_penalty
+
+
+class ImageProjectionFit(ImageFit):
+    """
+    Image fitting class that gets the beam size and location by independently fitting
+    the x/y projections. The default configuration uses a Gaussian fitting of the
+    profile with prior distributions placed on the model parameters.
+    """
+    projection_fit_method: Optional[MethodBase] = GaussianModel(use_priors=True)
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    def _fit_image(self, image: ndarray) -> dict:
+        x_projection = np.array(np.sum(image, axis=0))
+        y_projection = np.array(np.sum(image, axis=1))
+
+        proj_fit = ProjectionFit(model=self.projection_fit_method)
+
+        x_parameters = proj_fit.fit_projection(x_projection)
+        y_parameters = proj_fit.fit_projection(y_projection)
+
+        results = {
```

Comment:
I've mentioned this a couple times before, in this PR and in others, that I think we should move away from dictionaries of parameters. It's hard to validate what they actually contain as a user. It doesn't have to be fixed in this PR, I think we should come to a solution as a group.